src/analyses/ai_history.cpp

```diff
@@ -27,3 +27,169 @@ xmlt ai_history_baset::output_xml(void) const
   xml.data = out.str();
   return xml;
 }
+
+ai_history_baset::step_returnt
+call_stack_historyt::step(locationt to, const trace_sett &others) const
+{
+  DATA_INVARIANT(current_stack != nullptr, "current_stack must exist");
+
+  cse_ptrt next_stack = nullptr;
+
+  // First construct what the new history would be.
+  // This requires special handling if this edge is a function call or return
+  if(current_stack->current_location->is_function_call())
+  {
+    locationt l_return = std::next(current_stack->current_location);
+
+    if(l_return->location_number == to->location_number)
+    {
+      // Is skipping the function call, only update the location
+      next_stack =
+        cse_ptrt(new call_stack_entryt(l_return, current_stack->caller));
+    }
+    else
+    {
+      // An interprocedural (call) edge; add to the current stack
+      next_stack = cse_ptrt(new call_stack_entryt(to, current_stack));
+    }
+  }
+  else if(current_stack->current_location->is_end_function())
+  {
+    if(current_stack->caller == nullptr)
+    {
+      // Trying to return but there was no caller?
+      // Refuse to do the transition outright
```

Comment:
Why not assert it then?

Reply:
I did but there is a legitimate case when this happens -- the end of the initial calling function.  This is the code that finally finishes the analysis.